Java/parkservices/src/main/java/com/amazon/randomcutforest/parkservices/PredictorCorrector.java

```diff
@@ -408,152 +654,236 @@ protected <P extends AnomalyDescriptor> P detect(P result, IRCFComputeDescriptor
-            lastScore = score;
-            result.setAnomalyGrade(0);
-            return result;
+        result.setAnomalyGrade(workingGrade);
+        result.setThreshold(workingThreshold);
+        result.setInHighScoreRegion(inHighScoreRegion);
+
+        if (workingGrade > 0) {
+            if (correctedPoint != null) {
+                result.setExpectedRCFPoint(toDoubleArray(correctedPoint));
+            }
+            if (choice != originalChoice) {
+                attribution.renormalize(result.getRCFScore());
+            }
+            result.setAttribution(attribution);
+
+            result.setRelativeIndex(index);
         }
 
-        lastScore = score;
-        result.setAttribution(attribution);
-        result.setRelativeIndex(index);
+        saveScores(strategy, choice, scoreVector, correctedScore, result.transformMethod, shingleSize);
         return result;
     }
 
     public void setZfactor(double factor) {
-        thresholder.setZfactor(factor);
+        for (int i = 0; i < thresholders.length; i++) {
+            thresholders[i].setZfactor(factor);
+        }
     }
 
     public void setLowerThreshold(double lower) {
-        thresholder.setLowerThreshold(lower);
+        for (int i = 0; i < thresholders.length; i++) {
+            thresholders[i].setLowerThreshold(lower);
+        }
     }
 
     public void setScoreDifferencing(double persistence) {
-        thresholder.setScoreDifferencing(persistence);
+        for (int i = 0; i < thresholders.length; i++) {
+            thresholders[i].setScoreDifferencing(persistence);
+        }
     }
 
     public void setInitialThreshold(double initial) {
-        thresholder.setInitialThreshold(initial);
+        for (int i = 0; i < thresholders.length; i++) {
+            thresholders[i].setInitialThreshold(initial);
+        }
     }
 
-    public void setIgnoreNearExpectedFromAbove(double[] ignoreSimilarShift) {
-        if (ignoreSimilarShift != null) {
-            this.ignoreNearExpectedFromAbove = Arrays.copyOf(ignoreSimilarShift, ignoreSimilarShift.length);
+    public void setNumberOfAttributors(int numberOfAttributors) {
+        this.numberOfAttributors = numberOfAttributors;
+    }
+
+    public int getNumberOfAttributors() {
+        return numberOfAttributors;
+    }
+
+    public double[] getLastScore() {
+        return lastScore;
+    }
+
+    public void setLastScore(double[] score) {
+        if (score != null) {
+            System.arraycopy(score, 0, lastScore, 0, min(NUMBER_OF_MODES, score.length));
+        }
+    }
+
+    void validateIgnore(double[] shift) {
+        checkArgument(shift.length == 4 * baseDimension, () -> "has to be of length " + 4 * baseDimension);
+        for (double element : shift) {
+            checkArgument(element >= 0, "has to be non-negative");
         }
     }
 
-    public void setIgnoreNearExpectedFromBelow(double[] ignoreSimilarShift) {
+    public void setIgnoreNearExpected(double[] ignoreSimilarShift) {
```

Comment:
That approach would have 4 fields and inability to add any new type because of the byte 64 type and naming. By having one single array, it would be easier to add new options/ retire older options without changing the state classes.

Reply:
Understood this can be used in state classes. But I don't follow why we would do it in hresholdedRandomCutForest(Builder)

```
double[] nearExpected = new double[4 * base];
        builder.ignoreNearExpectedFromAbove.ifPresent(array -> {
            validateNonNegativeArray(array, base);
            System.arraycopy(array, 0, nearExpected, 0, base);
        });
        builder.ignoreNearExpectedFromBelow.ifPresent(array -> {
            validateNonNegativeArray(array, base);
            System.arraycopy(array, 0, nearExpected, base, base);
        });
        builder.ignoreNearExpectedFromAboveByRatio.ifPresent(array -> {
            validateNonNegativeArray(array, base);
            System.arraycopy(array, base, nearExpected, 2 * base, base);
        });
        builder.ignoreNearExpectedFromBelowByRatio.ifPresent(array -> {
            validateNonNegativeArray(array, base);
            System.arraycopy(array, base, nearExpected, 3 * base, base);
        });
        predictorCorrector.setIgnoreNearExpected(nearExpected);
```

We can assign to four specific fields directly without merging and splitting back.